### ISSUE TITLE
g1_to_block_last() can't deal with last G1 position

### ISSUE THREAD

**Issue by latk** [jeffreykegler/Marpa--R3#44]
When the `Context::g1_range()` function is invoked during the semantics of the last production in a parse, the end g1 position is not accepted by the recognizer's `g1_to_block_last()` method.
According to the docs, a “range” is inclusive:

> A range is an ordered set of contiguous locations specified by start location and end location: [S ... E]. […] A range is inclusive, so that it includes the location S as well as the location E. 
>
> – Marpa::R3::Recognizer, *Ranges*

So if `g1_range()` returns `(1, 2)`, then g1=2 is a valid location that should be resolvable to a physical location.

On the other hand, the docs for `g1_to_block_last()` do state:

> The block position returned by g1_to_block_last() will be an "actual" position, in the sense that there actually is a codepoint there.

Though I think returning the “end of block” position ought to be valid.

Background: when throwing errors from within the semantics, I want to include the offset of the _next_ input codepoint. In Marpa::R2, I had used: `my ($l0, $length) = $recce->g1_location_to_span($g1_end); return $l0 + $length;`

Code to reproduce:

``` perl
use Marpa::R3;  # master@a9103de
my $g = Marpa::R3::Grammar->new({
    source => \q{ A ::= 'a' B; B ::= 'b' action => main::semantics },
});
$g->parse(\"ab");

sub semantics {
  my $r = $Marpa::R3::Context::recognizer;
  my ($g1_start, $g1_end) = Marpa::R3::Context::g1_range();
  $r->g1_to_block_last($g1_end);
  return;
}
```

Error message:

```
============================================================
* THE MARPA SEMANTICS PRODUCED A FATAL ERROR
* THIS IS WHAT MARPA WAS DOING WHEN THE PROBLEM OCCURRED:
Computing value for rule: B ::= 'b'
* THIS WAS THE FATAL ERROR MESSAGE:
../kollos/kollos.lua:2729: slr:g1_to_block_last(2): bad argument
    Allowed values are 0-1

stack traceback:
        [C]: in function 'error'
        ../kollos/kollos.lua:3863: in function <../kollos/kollos.lua:3859>
        (...tail calls...)
Marpa::R3 exception at test.pl line 10.
Marpa::R3 exception at test.pl line 5.
Marpa::R3 exception at test.pl line 5.
```

Expected behaviour: no error.

As a variation of the problem, consider a grammar where the last symbol is nulling: `A ::= 'ab' B; B ::= action => main::semantics `. Now both `$g1_start` and `$g1_end` will refer to a G1 location that is rejected by `g1_to_block_last()`.

In all cases, `g1_to_block_first()` returns successfully.

**Comment by jeffreykegler** [comment]
Currently investigating this.  At this point it looks like both a documentation issue with G1 positions, and a bug in at least one of the routines that handles them.

**Comment by jeffreykegler** [comment]
Fixed, I hope, in the above referenced commit https://github.com/jeffreykegler/Marpa--R3/commit/995facec588e362107b2f5898842f944ae23cd6d  It was a documentation bug, and a bug in the 2nd return value of `g1_range()`. 

@latk Please verify and close if appropriate.

This fix was only to make the code&docs align with my current intentions for G1.  This may *not* address your use case.  If it does not, please open another issue.

**Comment by latk** [comment]
I updated Marpa::R3 but can still reproduce this exception. I updated the code & error message in the first message to represent the current state (this was only the Grammar class renaming, and the line number for the Lua error). 

Regardless of my use use case and what the correct behaviour would be, my code to reproduce passes my subjective “type check”, i.e. I get a G1 location from Marpa and then give it back. As this doesn't work, this indicates that either

 - the second return value of g1_range() is not always a valid G1 location, or
 - the argument to g1_block_to_last() only accepts a _subset_ of G1 locations.

Because the g1_range() return value is accepted by g1_block_to_first(), the error is likely in g1_block_to_last().

Interestingly, the lc_range() function seems to do almost exactly what I was trying to emulate, and works. As it is implemented within Lua, maybe the problem is on the Lua–Perl boundary? But the bindings for g1_block_to_{first,last}() do not seem to differ. But in the Lua code, I found that they differ in a bounds check:

``` lua
function _M.class_slr.g1_to_block_first(slr, g1_pos)
    local per_es = slr.per_es
    if g1_pos == #per_es then
        return slr:l0_current_pos()
    end
    local this_sweep = per_es[g1_pos+1]
    ...
```

versus

``` lua
function _M.class_slr.g1_to_block_last(slr, g1_pos)
    local per_es = slr.per_es
    local this_sweep = per_es[g1_pos+1]
    ...
```

That would conceivably be the reason why the last G1 location is accepted by g1_to_block_first(), but fails the later bounds check in g1_to_block_last().

**Comment by jeffreykegler** [comment]
I just now copied and pasted the code, which I take it has been edited, in the 1st message, and could not reproduce the error.  Perhaps one of us isn't capturing the version we thing we are.  I ran commit 89a87347c9fc980c5663f88a8b1801da796a4d13

I note the difference between `g1_block_to_{first,last}()`, but I think that means `g1_block_to_first()` is being too liberal in what it accepts.

Want to try again?
